kame/kame/libtest/libtest.1
.\"	$KAME: libtest.1,v 1.3 2000/11/09 03:04:08 itojun Exp $
.\" 
.\" Copyright (C) 1999 WIDE Project.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd October 26, 1999
.Dt LIBTEST 1
.Os KAME
.\"
.Sh NAME
.Nm libtest
.Nd test behavior of IPv6-related library
.\"
.Sh SYNOPSIS
.Nm
.\"
.Sh DESCRIPTION
.Nm
tests well known bugs in IPv6-realted libraries, like
.Xr inet_pton 3
prior to BIND 8.2.
.\"
.Sh RETURN VALUES
.Nm
exits with 0 on success, and non-zero on errors.
.\"
.Sh SEE ALSO
.Rs
.%A R. Gilligan
.%A S. Thomson
.%A J. Bound
.%A W. Stevens
.%T Basic Socket Interface Extensions for IPv6
.%D March 1999
.%N RFC2553
.Re
.\"
.Sh HISTORY
The
.Nm
command first appeared in WIDE/KAME IPv6 protocol stack kit.


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

kame/kame/libtest/libtest.c
/*
 * Copyright (C) 1999 WIDE Project.
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without loop prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
/*
 * $Id: libtest.c,v 1.8 2000/11/09 07:54:52 itojun Exp $
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/param.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <netdb.h>
#include <errno.h>
#include <err.h>

#include <netinet/in.h>
#include <arpa/inet.h>

int main __P((int, char **));
static void usage __P((void));
static int test_pton __P((void));
static int test_getnameinfo __P((void));
static int test_getnameinfo0 __P((const struct sockaddr *, const char *,
	const char *));

int
main(argc, argv)
	int argc;
	char **argv;
{
	int ch;
	extern int optind;
	int failure = 0;

	while ((ch = getopt(argc, argv, "")) != EOF) {
		switch (ch) {
		default:
			usage();
			exit(1);
		}
	}
	argc -= optind;
	argv += optind;

	if (argc != 0) {
		usage();
		exit(1);
	}

	failure += test_pton();
	failure += test_getnameinfo();

	printf("%d failure%s\n", failure, failure > 1 ? "s" : "");
	exit(failure);
}

static void
usage()
{
	fprintf(stderr, "usage: libtest\n");
}

static int
test_pton()
{
#define FUNCNAME	"test_pton"
	struct in6_addr a;
	int success = 0;

	printf("%s: start\n", FUNCNAME);

	/* test for broken inet_pton() (pre BIND82) */
	if (inet_pton(AF_INET6, "0:1:2:3:4:5:6:7:", &a) != 1)
		success++;
	else
		printf("%s: test 1 failed\n", FUNCNAME);
	if (inet_pton(AF_INET6, "0:1:2:3:4:5:6:7@", &a) != 1)
		success++;
	else
		printf("%s: test 2 failed\n", FUNCNAME);

	if (success != 2) {
		printf("the OS has broken inet_pton()\n");
		return 1;
	} else
		return 0;
#undef FUNCNAME
}

static int
test_getnameinfo()
{
#if 1
	struct sockaddr_storage ss;
#else
	struct {
		char buf[128];
	} ss;
#endif
	struct sockaddr *sa;
	struct sockaddr_in *sin;
	struct sockaddr_in6 *sin6;
	int failure;
	u_int8_t in6[16] = {
		0x10, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef,
		0x10, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef
	};

	failure = 0;
	sa = (struct sockaddr *)&ss;

	sin = (struct sockaddr_in *)&ss;
	memset(sin, 0, sizeof(*sin));
	sin->sin_len = sizeof(struct sockaddr_in);
	sin->sin_family = AF_INET;
	sin->sin_addr.s_addr = ntohl(0x7f000001);
	sin->sin_port = ntohs(9876);
	failure += test_getnameinfo0(sa, "127.0.0.1", "9876");

	sin6 = (struct sockaddr_in6 *)&ss;
	memset(sin6, 0, sizeof(*sin6));
	sin6->sin6_len = sizeof(struct sockaddr_in6);
	sin6->sin6_family = AF_INET6;
	memcpy(&sin6->sin6_addr, &in6, sizeof(sin6->sin6_addr));
	sin6->sin6_port = ntohs(9876);
	failure += test_getnameinfo0(sa,
	    "1023:4567:89ab:cdef:1023:4567:89ab:cdef", "9876");

	/* the rest of get{addr,name}info tests are in libinet6/test.c */

	return failure;
}

static int
test_getnameinfo0(sa, hans, pans)
	const struct sockaddr *sa;
	const char *hans;
	const char *pans;
{
#define FUNCNAME	"test_getnameinfo"
	char hbuf[NI_MAXHOST], pbuf[NI_MAXSERV];
	int i, j, k, l;
	int fail;
	int ntest;
#ifdef NI_WITHSCOPEID
	const int niflag = NI_NUMERICHOST | NI_NUMERICSERV | NI_WITHSCOPEID;
#else
	const int niflag = NI_NUMERICHOST | NI_NUMERICSERV;
#endif

	printf("%s: start addr %s port %s\n", FUNCNAME, hans, pans);

	fail = 0;
	ntest = 0;

	/*
	 * host/serv with NULL, or hostlen/servlen with 0, should not
	 * raise any error.
	 */
	for (i = 0; i < 3 * 3; i++) {
		/* the initialization here is for fool too-picky gcc setting */
		char *h = NULL, *p = NULL;
		int hl = 0, pl = 0;

		switch (i % 3) {
		case 0:	h = hbuf; hl = 0; break;
		case 1:	h = NULL; hl = sizeof(hbuf); break;
		case 2:	h = NULL; hl = 0; break;
		}
		switch ((i / 3) % 3) {
		case 0:	p = pbuf; pl = 0; break;
		case 1:	p = NULL; pl = sizeof(pbuf); break;
		case 2:	p = NULL; pl = 0; break;
		}
		ntest++;
		if (getnameinfo(sa, sa->sa_len, h, hl, p, pl, niflag) != 0) {
			printf("%s: test %d failed\n", FUNCNAME, ntest);
			fail++;
		}

	}

	/* success case: it should return correct result */
	ntest++;
	memset(hbuf, 0, sizeof(hbuf));
	memset(pbuf, 0, sizeof(pbuf));
	if (getnameinfo(sa, sa->sa_len, hbuf, sizeof(hbuf), pbuf, sizeof(pbuf),
	    niflag) == 0) {
		if (strcmp(hans, hbuf) == 0 && strcmp(pans, pbuf) == 0) {
			;
		} else {
			printf("%s: test %d failed with %s/%s\n", FUNCNAME,
				ntest, hbuf, pbuf);
		}
	} else {
		printf("%s: test %d failed\n", FUNCNAME, ntest);
	}

	/*
	 * RFC2553/X-open getnameinfo spec on short buffer.
	 * RFC2553 behavior - should raise error if the room is not enough.
	 * X/open behavior - caller must supply enough buffer.
	 */
	ntest++;
	j = k = l = 0;
	for (i = 1; i < strlen(hans); i++) {
		memset(hbuf, 0, sizeof(hbuf));
		if (getnameinfo(sa, sa->sa_len, hbuf, i, NULL, 0, niflag) != 0) {
			/* RFC2553 */
			j++;
		} else {
			if (strncmp(hans, hbuf, strlen(hbuf)) == 0) {
				/* X/open */
				k++;
			} else {
				printf("%s: test %d failed with len=%d\n",
					FUNCNAME, ntest, i);
				fail++;
			}
		}
		l++;
	}
	for (i = 1; i < strlen(pans); i++) {
		memset(pbuf, 0, sizeof(pbuf));
		if (getnameinfo(sa, sa->sa_len, NULL, 0, pbuf, i, niflag) != 0) {
			/* RFC2553 */
			j++;
		} else {
			if (strncmp(pans, pbuf, strlen(pbuf)) == 0) {
				/* X/open */
				k++;
			} else {
				printf("%s: test %d failed with len=%d\n",
					FUNCNAME, ntest, i);
				fail++;
			}
		}
		l++;
	}
	if (j == l && k == 0)
		printf("%s: RFC2553 behavior on short buffer\n", FUNCNAME);
	else if (j == 0 && k == l)
		printf("%s: X/open behavior on short buffer\n", FUNCNAME);
	else {
		printf("%s: test %d failed - random behavior (%d/%d)\n",
			FUNCNAME, ntest, j, k);
		fail++;
	}

	return fail ? 1 : 0;
#undef FUNCNAME
}
